File: 62989-.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <sstream>

using namespace std;

class column
{
public:
    column(string name, bool str) : name(name), is_string(str) {}
    string name;
    bool is_string;
};

class row
{
public:
    row(long long val) : i(val) {};
    row(string val) : str(val) {};
    long long i;
    string str;
};

class table
{
public:
    vector<column> colums;
    vector<vector<row>> rows;
};

map<string, bool> users;
map<string, table> tables;

int main()
{
    istringstream cline;
    string line, command, temp, table_name,
            column_name, user, what, role, type;

    getline(cin, line);
    cline.clear();
    cline.str(line);

    cline >> command;
    while (command != "done")
    {
        if (command == "create")
        {
            cline >> what;

            if (what == "user")
            {
                cline >> user >> role;
                users[user] = (role == "editor");
            }
            else if (what == "table")
            {
                cline >> table_name >> user;
                if (users[user])
                {
                    tables[table_name] = table();
                }
                else
                {
                    cout << "access denied" << endl;
                }
            }
        }
        else if (command == "delete")
        {
            cline >> temp >> table_name >> user;
            if (users[user])
            {
                tables.erase(table_name);
            }
            else
            {
                cout << "access denied" << endl;
            }
        }
        else if (command == "add")
        {
            cline >> what >> table_name;
            if (what == "column")
            {
                cline >> column_name >> type >> user;
                if (users[user])
                {
                    table & t = tables[table_name];
                    if (type == "int")
                    {
                        t.colums.emplace_back(column_name, false);
                        for (vector<row> & row_item : t.rows)
                        {
                            row_item.emplace_back(0);
                        }
                    }
                    else
                    {
                        t.colums.emplace_back(column_name, true);
                        for (vector<row> & row_item : t.rows)
                        {
                            row_item.emplace_back("null");
                        }
                    }
                }
                else
                {
                    cout << "access denied" << endl;
                }
            }
            else if (what == "row")
            {
                cline >> user;
                if (users[user])
                {
                    table & t = tables[table_name];
                    vector<row> new_row;
                    for (column & c : t.colums)
                    {
                        if (c.is_string)
                        {
                            new_row.emplace_back("null");
                        }
                        else
                        {
                            new_row.emplace_back(0);
                        }
                    }
                    t.rows.push_back(move(new_row));
                }
                else
                {
                    cout << "access denied" << endl;
                }
            }
        }
        else if (command == "remove")
        {
            cline >> what >> table_name;
            if (what == "column")
            {
                cline >> column_name >> user;
                if (users[user])
                {
                    int index = 0;
                    for (column & t : tables[table_name].colums)
                    {
                        if (t.name == column_name)
                        {
                            break;
                        }
                        else
                        {
                            ++index;
                        }
                    }

                    tables[table_name].colums.erase(tables[table_name].colums.begin() + index);
                    for (vector<row> & r : tables[table_name].rows)
                    {
                        r.erase(r.begin() + index);
                    }
                }
                else
                {
                    cout << "access denied" << endl;
                }
            }
            else if (what == "row")
            {
                int row_id;
                cline >> row_id >> user;

                if (users[user])
                {
                    tables[table_name].rows.erase(tables[table_name].rows.begin() + (row_id - 1));
                }
                else
                {
                    cout << "access denied" << endl;
                }
            }
        }
        else if (command == "change")
        {
            int row_id;
            cline >> table_name >> row_id >> column_name >> temp >> user;
            if (users[user])
            {
                table & t = tables[table_name];
                column * cl;
                int index = 0;
                for (column & c : t.colums)
                {
                    if (c.name == column_name)
                    {
                        cl = &c;
                        break;
                    }
                    else
                    {
                        ++index;
                    }
                }

                row & row_ = t.rows[row_id - 1][index];
                if (cl->is_string)
                {
                    row_.str = temp;
                }
                else
                {
                    row_.i = stoll(temp);
                }
            }
            else
            {
                cout << "access denied" << endl;
            }
        }
        else if (command == "print")
        {
            cline >> table_name;
            table & t = tables[table_name];
            vector<int> cols;

            vector<string> items;
            while (!cline.eof())
            {
                string s;
                cline >> s;
                items.push_back(move(s));
            }

            items.erase(items.end());
            if (items.size() == 1 && items[0] == "*")
            {
                int index = 0;
                for (column & c : t.colums)
                {
                    cols.push_back(index);
                    ++index;
                }
            }
            else
            {
                for (string c_name : items)
                {
                    int index = 0;
                    for (column & c : t.colums)
                    {
                        if (c.name == c_name)
                        {
                            cols.push_back(index);
                            break;
                        }
                        else
                        {
                            ++index;
                        }
                    }
                }
            }

            vector<column> & c = t.colums;
            for (vector<row> & r : t.rows)
            {
                for (int k = 0; k <cols.size(); ++k)
                {
                    int c_ = cols[k];

                    if (c[c_].is_string)
                    {
                        cout << r[c_].str << " ";
                    }
                    else
                    {
                        cout << r[c_].i << " ";
                    }
                }
                cout << endl;
            }
        }
        else if (command == "search")
        {
            cline >> table_name >> column_name >> temp >> user;

            table & t = tables[table_name];
            int ind = 0;
            bool is_str = false;
            vector<column> & cl = t.colums;
            for (column & c : t.colums)
            {
                if (c.name == column_name)
                {
                    is_str = c.is_string;
                    break;
                }
                else
                {
                    ++ind;
                }
            }

            for (vector<row> & r : t.rows)
            {
                if (is_str)
                {
                    if (r[ind].str != temp)
                    {
                        continue;
                    }
                }
                else
                {
                    if (r[ind].i != stoll(temp))
                    {
                        continue;
                    }
                }

                for (int k = 0; k < r.size(); ++k)
                {
                    if (cl[k].is_string)
                    {
                        cout << r[k].str << " ";
                    }
                    else
                    {
                        cout << r[k].i << " ";
                    }
                }
                cout << endl;
            }
        }

        getline(cin, line);
        cline.clear();
        cline.str(line);

        cline >> command;
    }

    return 0;
}

File: 62989.py
import sqlite3
db = sqlite3.connect(":memory:")
types = {"int": "INTEGER  NOT NULL default 0", "string": "TEXT NOT NULL default `null`"}
users = {}
tables = {}
insert_id = 0

while True:
    q = input().split()
    if q[0] == "done":
        break;

    elif q[0] == "create":
        if q[1] == "user":
            users[q[2]] = (q[3] == "editor")
        elif q[1] == "table":
            if users[q[-1]]:
                db.execute("create table `%s` (h_id integer)" % q[2])
                tables[q[2]] = ["h_id"]
            else:
                print("access denied")

    elif q[0] == "delete":
        if users[q[-1]]:
            db.execute("drop table `%s`" % q[2])
        else:
            print("access denied")

    elif q[0] == "add":
        if users[q[-1]] == False:
            print("access denied")
        else:
            if q[1] == "column":
                db.execute("alter table `%s` add `%s` %s" % (q[2], q[3], types[q[4]]))
                tables[q[2]].append(q[3])
            elif q[1] == "row":
                db.execute("insert into `%s`(h_id) values(%d)" % (q[2], insert_id))
                insert_id += 1

    elif q[0] == "remove":
        if users[q[-1]] == False:
            print("access denied")
        else:
            if q[1] == "column":
                c = db.execute("select * from sqlite_master where tbl_name = '%s'" % q[2])
                for i in c:
                    c = i[-1]
                    break
                c = c.replace("`%s`" % q[2], "`%s_old`" % q[2]).split(',')
                for i in range(len(c)):
                    if q[3] in c[i]:
                        c.pop(i)
                        break
                c = ",".join(c)
                if c[-1] != ')':
                    c += ')'

                db.execute(c)
                tables[q[2]].remove(q[3])
                cl = ", ".join(tables[q[2]])
                quer = "insert into `%s_old` (%s) select %s from `%s`" % (q[2], cl, cl, q[2])
                db.execute(quer)
                db.execute("drop table `%s`" % q[2])
                db.execute("alter table `%s_old` RENAME TO `%s`" % (q[2], q[2]))
            elif q[1] == "row":
                db.execute("delete from `%s` where h_id in (select h_id from `%s` limit 1 offset %d)" % (q[2], q[2], int(q[3]) - 1))
            
    elif q[0] == "change":
        if users[q[-1]] == False:
            print("access denied")
        else:
            db.execute("update `%s` set `%s` = '%s' where h_id in (select h_id from `%s` limit 1 offset %d)" % (q[1], q[3], q[4], q[1], int(q[2]) - 1))

    elif q[0] == "search":
        for i in db.execute("select * from `%s` where %s = '%s'" % (q[1], q[2], q[3])):
            for j in i[1:]:
                print(j, end=" ")
            print()

    elif q[0] == "print":
        if q[2] == "*":
            for i in db.execute("select * from `%s`" % (q[1])):
                for j in i[1:]:
                    print(j, end=" ")
                print()
        else:
            cl = q[2:-1]
            quer = "select h_id, %s from `%s` order by " % (",".join(cl), q[1])
            for c in cl:
                quer += "%s asc, " % c
            quer += "h_id asc"
            for i in db.execute(quer):
                for j in i[1:]:
                    print(j, end=" ")
                print()
            
